File: 875.KokoEatingBananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1, right = 1;
        for (int p : piles){
            right = Math.max(right, p);
        }
        int res = right;
        while (left < right){
            int mid = (left + right) / 2;
            int hoursSpent = 0;
            for ( int p : piles){
                hoursSpent += Math.ceil((double) p / mid);
            }

            if(hoursSpent <= h){
                res = Math.min(res, mid);
                right = mid;
            }
            else{
                left = mid + 1;
            }
        }
        return res;
    }
}

File: 875.KokoEatingBananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        totalTime = max(piles)
        while l <= r:
            k = (l + r) // 2
            cnt = 0
            for i in range(len(piles)):
                cnt += math.ceil(piles[i] / k)
            if cnt > h:
                l = k + 1
            else:
                totalTime = min(totalTime, k)
                r = k - 1
        return totalTime
    
